### DISCUSSION TITLE
Arduino Nano 33 BLE hanging when using RFM96 for AX.25 AFSK

### DISCUSSION THREAD

**Discussion by elaidlaw** [jgromes/RadioLib#407] [General]
I'm using an Arduino Nano 33 BLE and RFM96 to send AX.25 packets with AFSK, but after several packets have been sent, the Arduino becomes unresponsive and I have to manually reset it to its bootloader and reupload the sketch before it will work again.

It always fails at the same place in transmission:
```
[AX.25] Sending UI frame ... R	1	9	
R	1	9	
R	1	9	
W	1	9	
R	1	9	
R	40	0	
W	40	0	
R	40	0	
R	1	9	
R	D	18	
W	D	18	
R	D	18	
R	31	0	
W	31	0	
R	31	0	
R	1	9	
R	1	9	
W	1	B	
R	1	B	
R	1	B	
R	1	B	
W	1	9	
R	1	B	
R	1	B	
R	1	9	
[AX.25] Sending UI frame ... R	1	9	
R	1	9	
R	1	9	
W	1	9	
R	1	9	
R	40	0	
W	40	0	
R	40	0	
R	1	9	
R	D	18	
W	D	18	
R	D	18	
R	31	0	
W	31	0	
R	31	0	
R	1	9	
R	1	9	
W	1	B	
R	1	B
```

Any ideas what might be happening? I saw another post on here about a similar issue and it sounds like you don't have a Nano BLE to try it out on, but if it would help I would be happy to cover the cost for one.

Another thing to note is that we're not actually receiving the packets transmitted using the Nano BLE, whereas we are when using an Uno. Even when we use the Uno though, we're only getting some of the packets. I'm sending the string `"Hello Anthony! This is test message 4 to see if the radio works. Does it work?"` 6 times, replacing the number 4 here with 0-5 for each packet respectively. What's super weird is that we consistently receive the same two packets every time the set of six are sent, and none of the others.

So two questions: any idea why this happens? (it's possible our receiver has a problem so if you think that's the case just say so) And then why would it not work on the Nano BLE at all?

**Comment by jgromes** [comment]
Freezing after a couple messages were sent seems like a memory issue. Does it only happen with AFSK, or is 2-FSK the same? And does it still happen if you enable `RADIOLIB_STATIC_ONLY` in BuildOpt.h?

> we're not actually receiving the packets transmitted using the Nano BLE

I'm not sure what is the exact setup. Are you transmitting FSK from Uno and not receiving it on Nano BLE? Could you post the code?

**Comment by elaidlaw** [comment]
I'm using a very slightly modified version of the AX.25 Transmit AFSK example. We're receiving using a software-defined radio that just plugs into the computer via USB. When we use this code on the Arduino Uno, we get the issue where only some of the packets are received. When we use this code on the Arduino Nano 33 BLE, none of the packets are received.

```c++
/*
   RadioLib AX.25 Transmit AFSK Example

   This example sends AX.25 messages using
   SX1278's FSK modem. The data is modulated
   as AFSK at 1200 baud using Bell 202 tones.

   Other modules that can be used for AX.25
   with AFSK modulation:
    - SX127x/RFM9x
    - RF69
    - SX1231
    - CC1101
    - nRF24
    - Si443x/RFM2x

   For default module settings, see the wiki page
   https://github.com/jgromes/RadioLib/wiki/Default-configuration

   For full API reference, see the GitHub Pages
   https://jgromes.github.io/RadioLib/
*/

// include the library
#include <RadioLib.h>

// SX1278 has the following connections:
// NSS pin:   10
// DIO0 pin:  2
// RESET pin: 9
// DIO1 pin:  3
SX1278 radio = new Module(4, 3, 2, 5);

// or using RadioShield
// https://github.com/jgromes/RadioShield
//SX1278 radio = RadioShield.ModuleA;

// create AFSK client instance using the FSK module
// pin 5 is connected to SX1278 DIO2
AFSKClient audio(&radio, 6);
//
//// create AX.25 client instance using the AFSK instance
AX25Client ax25(&audio);

void setup() {
  Serial.begin(9600);

  // initialize SX1278 with default settings
  Serial.print(F("[SX1278] Initializing ... "));
  int state = radio.beginFSK(435.0, 1.2, 125.0, 10, 16, false);
  radio.setOutputPower(20);
  radio.variablePacketLengthMode(SX127X_MAX_PACKET_LENGTH);
  uint8_t syncWord[] = {0x12, 0xAD};
  radio.setSyncWord(syncWord, 2);
  radio.setDataShaping(RADIOLIB_SHAPING_NONE);
  radio.setCurrentLimit(60);
  radio.setEncoding(RADIOLIB_ENCODING_NRZ);
  radio.setCRC(true);
  radio.disableAddressFiltering();
  radio.setRSSIConfig(2, 0);
//
  // when using one of the non-LoRa modules for AX.25
  // (RF69, CC1101,, Si4432 etc.), use the basic begin() method
  // int state = radio.begin();

  if(state == ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while(true);
  }

  // initialize AX.25 client
  Serial.print(F("[AX.25] Initializing ... "));
  // source station callsign:     "N7LEM"
  // source station SSID:         0
  // preamble length:             8 bytes
  state = ax25.begin("EVESF");
  if(state == ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while(true);
  }

  // Sometimes, it may be required to adjust audio
  // frequencies to match the expected 1200/2200 Hz tones.
  // The following method will offset mark frequency by
  // 100 Hz up and space frequency by 100 Hz down
  /*
    Serial.print(F("[AX.25] Setting correction ... "));
    state = ax25.setCorrection(100, -100);
    if(state == ERR_NONE) {
      Serial.println(F("success!"));
    } else {
      Serial.print(F("failed, code "));
      Serial.println(state);
      while(true);
    }
  */
}

void loop() {
//  // send AX.25 unnumbered information frame
  delay(5000);
  Serial.print(F("[AX.25] Sending UI frame ... "));
//  // destination station callsign:     "NJ7P"
//  // destination station SSID:         0
  int state = ax25.transmit("Hello Anthony! This is test message 0 to see if the radio works. Does it work?", "NJ7P");
  delay(100);
  state = ax25.transmit("Hello Anthony! This is test message 1 to see if the radio works. Does it work?", "NJ7P");
  delay(100);
  state = ax25.transmit("Hello Anthony!?", "NJ7P");
  delay(100);
  state = ax25.transmit("11111111", "EEFD");
  delay(100);
  state = ax25.transmit("Hello Anthony! This is test message 4 to see if the radio works. Does it work?", "NJ7P");
  delay(100);
  state = ax25.transmit("Hello Anthony! This is test message 3 to see if the radio works. Does it work?", "NJ7P");
  delay(100);
//  if (state == ERR_NONE) {
//    // the packet was successfully transmitted
//    Serial.println(F("success!"));
//
//  } else {
//    // some error occurred
//    Serial.print(F("failed, code "));
//    Serial.println(state);
//
//  }
//
  delay(5000);
}
```

I'm gonna try the static memory thing now.

**Reply by jgromes** [reply]
What software are you using for the reception? I usually test AX.25 with soundmodem. On a side note, the only SX1278 parameter you have to set is the frequency. `ax25.begin()` will take care of the rest.

Also, did you try some frequency corrections?

**Reply by elaidlaw** [reply]
We actually are using a soundmodem.

From the guy who knows more about radios than I do:

The soundmodem data waterfall shows that the AFSK tones received from the radio are outside the nominal 1200 and 2200 Hz of AFSK (say, closer to 1000Hz and 2400Hz). Could this affect the reception of some specific messages?

^ is this related to the frequency correction question? We have adjusted the frequency by a few kilohertz to match what it seems like the radio is actually outputting if that's what you mean.

**Reply by jgromes** [reply]
What I meant is the following: If your audio tones are at 1000Hz and 2400 Hz instead of 1200 and 2200, then you should do `ax25.setCorrection(200, -200);`, which should offset the low tone by +200 Hz and the high tone by -200 Hz.

**Reply by kazar4** [reply]
Hi, I am also working on the team, and today we tried the setCorrection(), we looked at the audio tones and saw they were about 100 off so we added ax25.setCorrection(100, -100); but it did not work even though we could see the audio tones on the correct spot. 

Also  we found an issue with the RFM95 (not 96) where we got status code 12 after setting the frequency to 868 to 915, saying that the module does not support the frequency. We used RFM95 as the type for our radio variable. We are curious as to why this error happened since the module (https://www.amazon.com/gp/product/B01HYO608A/ref=ppx_yo_dt_b_asin_title_o00_s01?ie=UTF8&psc=1) supports those two frequencies (it ran with 415 as the frequency fine)

**Reply by kazar4** [reply]
We found a fix for the RFM95, it seems to be inheriting from the SX1278 module, and so when we comment out 
RADIOLIB_CHECK_RANGE(freq, 137.0, 525.0, ERR_INVALID_FREQUENCY);
in SX1278.cpp
It transmits successfully. Our guess is that it is inheriting the check_range from SX1278 before getting to the check_range in the RFM95.cpp class which has the correct bounds shown here
RADIOLIB_CHECK_RANGE(freq, 862.0, 1020.0, ERR_INVALID_FREQUENCY);

**Comment by elaidlaw** [comment]
Getting a couple errors compiling with `RADIOLIB_STATIC_ONLY` set:

```
C:\Users\laidl\Documents\Arduino\libraries\RadioLib\src\protocols\RTTY\RTTY.cpp: In constructor 'ITA2String::ITA2String(char)':
C:\Users\laidl\Documents\Arduino\libraries\RadioLib\src\protocols\RTTY\RTTY.cpp:6:20: error: incompatible types in assignment of 'char*' to 'char [256]'
   _str = new char[1];
                    ^
C:\Users\laidl\Documents\Arduino\libraries\RadioLib\src\protocols\RTTY\RTTY.cpp: In constructor 'ITA2String::ITA2String(const char*)':
C:\Users\laidl\Documents\Arduino\libraries\RadioLib\src\protocols\RTTY\RTTY.cpp:13:27: error: incompatible types in assignment of 'char*' to 'char [256]'
   _str = new char[_len + 1];
                           ^
```

**Reply by jgromes** [reply]
Oops, sorry about that. Haven't used static only mode in a while - it's fixed now.

**Comment by elaidlaw** [comment]
I tested with `RADIOLIB_STATIC_ONLY` and got the same issue where it transmits a few times and then hangs. I can't compile with verbose output because of another compilation error so I can't give any debug output. The problem seems slightly less severe though (?) because I can reset the Nano BLE with its reset button and it works again. Any other ideas of what I can do?

**Reply by jgromes** [reply]
I can try on a different platform, but I've been using AX.25 on Arduino Uno running for several hours at a time, without issues.

> I can't compile with verbose output because of another compilation error

Could you please post that error? I don't seem to be able to replicate that.

**Reply by kazar4** [reply]
Hello, the google doc below has the error we got for verbose mode along with a general overview of the current problems we are experiencing [(google doc link)](https://docs.google.com/document/d/14yu-JqycRXVyqGZNyV-YdxhmDCCoOi3dk-NDju_bbok/edit?usp=sharing)

I would just like to add that these issues seem to be Nano BLE specific, as on an Arduino Uno all our code works without hanging and gets decoded, and so we would be more than happy to send you one if you think it would help. 

However we understand how much time this kind of thing can take, so once again thank you for everything. I hope you have a good week.

**Reply by jgromes** [reply]
Here's the gist of the document for future reference (once the document the link points to is inevitably lost to the void in a couple of years).

Sketch:
```c++
/*
   RadioLib AX.25 Transmit AFSK Example
   This example sends AX.25 messages using
   SX1278's FSK modem. The data is modulated
   as AFSK at 1200 baud using Bell 202 tones.
   Other modules that can be used for AX.25
   with AFSK modulation:
    - SX127x/RFM9x
    - RF69
    - SX1231
    - CC1101
    - nRF24
    - Si443x/RFM2x
   For default module settings, see the wiki page
   https://github.com/jgromes/RadioLib/wiki/Default-configuration
   For full API reference, see the GitHub Pages
   https://jgromes.github.io/RadioLib/
*/

// include the library
#include <RadioLib.h>

// SX1278 has the following connections:
// NSS pin:   10
// DIO0 pin:  2
// RESET pin: 9
// DIO1 pin:  3
RFM96 radio = new Module(5, 2, 6, 3);
int count = 0;

// or using RadioShield
// https://github.com/jgromes/RadioShield
//SX1278 radio = RadioShield.ModuleA;

// create AFSK client instance using the FSK module
// pin 5 is connected to SX1278 DIO2
AFSKClient audio(&radio, 4);

// create AX.25 client instance using the AFSK instance
AX25Client ax25(&audio);

void setup() {
  Serial.begin(9600);

  // initialize SX1278 with default settings
  Serial.print(F("[SX1278] Initializing ... "));
  int state = radio.beginFSK(435.0, 1.2, 125.0, 10, 16, false);
  //int state = radio.beginFSK(435.0);
  radio.setOutputPower(17);
  radio.setCurrentLimit(140);

  // when using one of the non-LoRa modules for AX.25
  // (RF69, CC1101,, Si4432 etc.), use the basic begin() method
  // int state = radio.begin();

  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while(true);
  }

  // initialize AX.25 client
  Serial.print(F("[AX.25] Initializing ... "));
  // source station callsign:     "N7LEM"
  // source station SSID:         0
  // preamble length:             8 bytes
  state = ax25.begin("N7LEM");
  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while(true);
  }
  ax25.setCorrection(200, -200);

  // Sometimes, it may be required to adjust audio
  // frequencies to match the expected 1200/2200 Hz tones.
  // The following method will offset mark frequency by
  // 100 Hz up and space frequency by 100 Hz down
  /*
    Serial.print(F("[AX.25] Setting correction ... "));
    state = ax25.setCorrection(100, -100);
    if(state == RADIOLIB_ERR_NONE) {
      Serial.println(F("success!"));
    } else {
      Serial.print(F("failed, code "));
      Serial.println(state);
      while(true);
    }
  */
}

void loop() {
  // send AX.25 unnumbered information frame
  Serial.print(F("[AX.25] Sending UI frame ... "));
  // destination station callsign:     "NJ7P"
  // destination station SSID:         0
  String myString = String(count);
  String str = "message 1 try" + myString;
  Serial.println(str);
  count = count + 1;
  int state = ax25.transmit(str.c_str(), "NJ7P");
  if (state == RADIOLIB_ERR_NONE) {
    // the packet was successfully transmitted
    Serial.println(F("success!"));
    #ifdef RADIOLIB_STATIC_ONLY
    Serial.println("static");
    #endif

  } else {
    // some error occurred
    Serial.print(F("failed, code "));
    Serial.println(state);

  }

  delay(3000);
}
```

The main issue is that after 37 transmissions, the Arduino hangs. Secondary issue with verbose mode giving error should be fixed now.

The fact that the radio module continues to output unmodulated carrier basically just means that the Arduino stopped controlling it sometimes during the transmission. That it always stops after 37 seconds would indicate a memory issue (although the fact that it persists in static mode seems to contradict that). Still, could you try to measure the free memory during runtime?

Soundmodem not being able to decode the packets might be due to issues in implementation of `tone()` on Arduino Nano 33 BLE, since that's what is internally used to generate the Bell 202 tones for AX.25.

**Reply by jgromes** [reply]
@kazar4 if you have an overabundance of Arduino Nano 33 BLEs, get in touch on gromes.jan@gmail.com and I'll happily try it with actual hardware ;)

**Reply by kazar4** [reply]
Here is the output with heap size, not too familiar on printing free memory so if this isnt the correct information I can try other method
[memoryData.txt](https://github.com/jgromes/RadioLib/files/7629646/memoryData.txt)
Also here is a pdf version of the document I linked in my last reply
[Arduino Nano BLE RadioLib Testing.pdf](https://github.com/jgromes/RadioLib/files/7629795/Arduino.Nano.BLE.RadioLib.Testing.pdf)


The code used was this (the process of getting memory seems to be different for the ble)
```c++ /*
   RadioLib AX.25 Transmit AFSK Example
   This example sends AX.25 messages using
   SX1278's FSK modem. The data is modulated
   as AFSK at 1200 baud using Bell 202 tones.
   Other modules that can be used for AX.25
   with AFSK modulation:
    - SX127x/RFM9x
    - RF69
    - SX1231
    - CC1101
    - nRF24
    - Si443x/RFM2x
   For default module settings, see the wiki page
   https://github.com/jgromes/RadioLib/wiki/Default-configuration
   For full API reference, see the GitHub Pages
   https://jgromes.github.io/RadioLib/
*/

// include the library
#include <RadioLib.h>
#include "mbed.h"
#include "mbed_mem_trace.h"

// SX1278 has the following connections:
// NSS pin:   5
// DIO0 pin:  2
// RESET pin: 6
// DIO1 pin:  3
RFM95 radio = new Module(5, 2, 6, 3);
int count = 0;

// or using RadioShield
// https://github.com/jgromes/RadioShield
//SX1278 radio = RadioShield.ModuleA;

// create AFSK client instance using the FSK module
// pin 5 is connected to SX1278 DIO2
AFSKClient audio(&radio, 4);

// create AX.25 client instance using the AFSK instance
AX25Client ax25(&audio);

void print_memory_info(char* printEvent, int iSize) {
    // allocate enough room for every thread's stack statistics
    int cnt = osThreadGetCount();
    mbed_stats_stack_t *stats = (mbed_stats_stack_t*) malloc(cnt * sizeof(mbed_stats_stack_t));
 
    cnt = mbed_stats_stack_get_each(stats, cnt);
    for (int i = 0; i < cnt; i++) {
        snprintf_P(printEvent, iSize, "Thread: 0x%lX, Stack size: %lu / %lu\r\n", stats[i].thread_id, stats[i].max_size, stats[i].reserved_size);
        Serial.println(printEvent);
    }
    free(stats);
 
    // Grab the heap statistics
    mbed_stats_heap_t heap_stats;
    mbed_stats_heap_get(&heap_stats);
    snprintf_P(printEvent, iSize, "Heap size: %lu / %lu bytes\r\n", heap_stats.current_size, heap_stats.reserved_size);
    Serial.println(printEvent);
}

void setup() {
  Serial.begin(9600);

  // initialize SX1278 with default settings
  Serial.print(F("[SX1278] Initializing ... "));
  int state = radio.beginFSK(915.0, 1.2, 125.0, 10, 16, false);
  radio.setOutputPower(17);
  radio.setCurrentLimit(140);

  // when using one of the non-LoRa modules for AX.25
  // (RF69, CC1101,, Si4432 etc.), use the basic begin() method
  // int state = radio.begin();

  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while(true);
  }

  // initialize AX.25 client
  Serial.print(F("[AX.25] Initializing ... "));
  // source station callsign:     "N7LEM"
  // source station SSID:         0
  // preamble length:             8 bytes
  state = ax25.begin("N7LEM");
  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while(true);
  }
  ax25.setCorrection(200, -200);

  // Sometimes, it may be required to adjust audio
  // frequencies to match the expected 1200/2200 Hz tones.
  // The following method will offset mark frequency by
  // 100 Hz up and space frequency by 100 Hz down
  /*
    Serial.print(F("[AX.25] Setting correction ... "));
    state = ax25.setCorrection(100, -100);
    if(state == RADIOLIB_ERR_NONE) {
      Serial.println(F("success!"));
    } else {
      Serial.print(F("failed, code "));
      Serial.println(state);
      while(true);
    }
  */
}

void loop() {

  char buffer [1000];
  print_memory_info(buffer, 1000);
  
  // send AX.25 unnumbered information frame
  Serial.print(F("[AX.25] Sending UI frame ... "));
  // destination station callsign:     "NJ7P"
  // destination station SSID:         0
  String myString = String(count);
  String str = "message 1 try" + myString;
  Serial.println(str);
  count = count + 1;
  int state = ax25.transmit(str.c_str(), "NJ7P");
  if (state == RADIOLIB_ERR_NONE) {
    // the packet was successfully transmitted
    Serial.println(F("success!"));
    #ifdef RADIOLIB_STATIC_ONLY
    Serial.println("static");
    #endif

  } else {
    // some error occurred
    Serial.print(F("failed, code "));
    Serial.println(state);

  }

  delay(300);
}

**Comment by elaidlaw** [comment]
As a side note, does RadioLib support receiving AX.25 over AFSK? Looks like receiving support is only there for FSK and LoRa?

**Reply by jgromes** [reply]
Receiving AFSK is not supported (I'm not even sure it can be done due to very hacky way AFSK is implemented). Also, while you can receive AX.25 frames sent with 2-FSK or LoRa, decoding them is also not supported, because I haven't yet had the time to implement that.

**Comment by elaidlaw** [comment]
For the issue where we're only getting some of the messages, this is what we're seeing ([image](https://drive.google.com/file/d/1xg39xM50O_0kY7LDJvnzo5bNyrLUrV6L/view?usp=sharing)). We're trying to send "message 1 tryX" where X counts up from 0, but as you can see we're only getting some of the messages. The ones we're getting are consistent if we re-run the program as well.

**Reply by jgromes** [reply]
It looks like all packets are getting transmitted, I think we can rule out memory issues.

Actually, I think I might have seen this in the past, I recall some packets were unable to be decoded by soundmodem. I think there might be some encoding issue in RadioLib implementation of AX.25, which makes some packets undecodable. Unfortunately I haven't had the time to track it down yet.

**Reply by kazar4** [reply]
Just curious if it would be possible to look into it at some point. We are trying now to test with LoRa to see if that helps with the issue, but in terms of radios that will work and things we have already tried there seems to not be many other options other than AX.25.

No worries otherwise.

**Reply by jgromes** [reply]
The encoding issue is something that I am looking into, unfortunately without any luck so far - but a PR that fixes it would be very welcome ;)

**Comment by jgromes** [comment]
@kazar4 Finally managed to track it down, so the encoding issue should now be fixed.

@elaidlaw I tried the original sketch you posted in this discussion and all packets are decoded by soundmodem correctly.

**Reply by elaidlaw** [reply]
Thank you so much, it looks like it's working now!

**Comment by jgromes** [comment]
@kazar4 @elaidlaw So with the Nano BLE, I managed to track down the cause of this hanging. Unfortunately I'm not sure if I will be able to do anything about it.

For starters, here's a very simple code snippet demontrating the issue - notice it is not using RadioLib code at all.

```c++
pinMode(6, OUTPUT);
while(true) {
  tone(6, 2200);
  delayMicroseconds(833);
  tone(6, 1200);
  delayMicroseconds(833);
}
```

In theory, this should indefinitely flip between outputting a 1200 Hz and 2200 Hz tone on the pin 6. In reality, this will only work for a couple seconds, and then the Arduino will hang. That does not happen on Uno.

The reason is that on Neno 33 BLE, `tone()` is implemented in a pretty inefficient way IMHO. Here's the implementation from Tone.cpp:

```c++
Tone* active_tone = NULL;

void tone(uint8_t pin, unsigned int frequency, unsigned long duration) {
	if (active_tone) {
		delete active_tone;
	}
	Tone* t = new Tone(digitalPinToPinName(pin), frequency, duration);
	t->start();
	active_tone = t;
};

void noTone(uint8_t pin) {
	if (active_tone) {
		active_tone->stop();
		delete active_tone;
		active_tone = NULL;
	}
};
```

Every time you call `tone()`, a new instance of the `Tone` class is created using dynamic memory allocation (using the `new` operator). Dynamic memory allocation takes a long time and may not always succeed. The `tone` implementation doesn't check that, and if allocation fails, it will attempt to derefernce a null pointer `t`, which will hang or crash. So if you call `tone` enough times, it will crash.

Unfortunately this is not only the case for Nano 33 BLE, but any mbed OS Arduino (e.g. RP2040 or Portenta).

I'm going to try a couple things to get it working reliably, but most likely it will involve implementing RadioLib's own `tone`, just at is done for ESP32.

**Reply by jgromes** [reply]
Now that's better: 

```
1:Fm N7LEM To NJ7P <UI R Pid=F0 Len=16> [17:03:40R] [--+]
message 1 try344
1:Fm N7LEM To NJ7P <UI R Pid=F0 Len=16> [17:03:43R] [--+]
message 1 try345
1:Fm N7LEM To NJ7P <UI R Pid=F0 Len=16> [17:03:46R] [--+]
message 1 try346
1:Fm N7LEM To NJ7P <UI R Pid=F0 Len=16> [17:03:49R] [--+]
message 1 try347
```

Running your test sketch that previously failed after 37 messages.

I wrote my own implementation of tone using PWM. The fix was pushed and is applied to all Arduino mbed boards, including Nano 33 BLE.

**Reply by kazar4** [reply]
Sorry for the late reply but thank you so much, word do not describe how much this helps us!

This has been one of the largest issues we have been worried about on the project and so getting this fixed is really nice to see.
In terms of testing on our side the fixes do seem to be working in all our use cases. One thing if you ever have the time that we were a little worried about is when sending messages with AX.25 from the BLE we drop more messages than with an Uno. By dropping messages I mean we see the waveform on SDRConsole but they do not get demodulated on soundmodem. So if you ever get a chance to check if you get a similar demodulating issue we would really appreciate it (and if you do we are curious what radio you are using). Overall though unlike before it does not seem deterministic unlike before (where the same message would not send), because that atleast means we can send the same message multiple times for redundancy and one should get through.

A seperate issue that Isn't affecting us but it may for others in the future is using the RFM95, when running the code
`RFM95 radio = new Module(5, 2, 6, 3);`
and then doing `radio.beginFSK(915.0, 1.2, 125.0, 10, 16, false);` we get error code -12
We looked into the code and found that RFM95 inherits the SX1278 and so we found
`RADIOLIB_CHECK_RANGE(freq, 137.0, 525.0, ERR_INVALID_FREQUENCY);`
in `SX1278.cpp` to be a problem as even though the RFM95 is specifically rated for 915, this line will make it fail (it works fine when commented out). I would of made a Pull Request but I am not familiar with C++ inheritance. But since we are not using that radio anymore this is not an issue for us, I just figured id mention it.

**Reply by jgromes** [reply]
Glad it's working now! Out of curiousity, what's the project? Feel free to share it in the [show and tell](https://github.com/jgromes/RadioLib/discussions/categories/show-and-tell) category, if you can ;)

I'll try to play around with AX.25 on the Nano 33 a bit more. Let me know if you notice a recurring pattern in your usage.

The issue with RFM95 should also be fixed. In the future, feel free to report that in a separate issue.

**Reply by kazar4** [reply]
Thank you!

We are building a cube satellite with plans to launch early 2022!

Here is a link If you are interested [https://sbudnic.space](https://sbudnic.space), some of the parts of the website are not updated but will be soon (for example back when the website was made we were trying a different radio)

**Reply by jgromes** [reply]
Awesome! Good luck to you guys, I'll keep an eye out on the project.

Another satellite flying RadioLib, that's always nice to see - I should start writing down the flight heritage ;)

**Reply by kazar4** [reply]
When you say another satellite flying RadioLib, do you happen to know the name or contact of that other satellite? Just curious as it would be helpful if we could ask them about their system (Like if they used LoRa for transmission).

**Reply by jgromes** [reply]
FOSSASAT-1, 1B and 2 used RadioLib, I was lead software on those. FOSSASAT-1 was launched in December 2019, but failed to deploy antennas and solar panels. It used SX1268, transmitting LoRa, FSK and RTTY. Even with undeployed antenna we were able to get some information from the RTTY beacon, and there was also a couple LoRa messages retrieved by recording and replying from SDR. Due to the undeployed solar panels it also had a slightly negative power budget and died about 3 months later.

FOSSASAT-1B and 2 were lost in Firefly Apha launch failure this September. There's a second FOSSASAT-1B that will be flying with Firefly in Q1-Q2 2022. All of these are open-source on GitHub (https://github.com/FOSSASystems).

Then there was the BeliefSat project (https://github.com/NewLeapKjsieit/BeliefSat), but I'm not sure what's the status there.

Those are the ones I know of, but I've been told on a couple of occasions that RadioLib is being used a lot by amateur satellites and HABs, because it supports a lot of digimodes on the most common radio modules, so there might be more.
